lib/spack/spack/environment.py

```diff
@@ -1509,6 +1509,66 @@ def concretized_specs(self):
         for s, h in zip(self.concretized_user_specs, self.concretized_order):
             yield (s, self.specs_by_hash[h])
 
+    def matching_spec(self, spec):
+        """
+        Given a spec (likely not concretized), find a matching concretized
+        spec in the environment.
+
+        The matching spec does not have to be installed in the environment,
+        but must be concrete (specs added with `spack add` without an
+        intervening `spack concretize` will not be matched).
+
+        If there is a single root spec that matches the provided spec or a
+        single dependency spec that matches the provided spec, then the
+        concretized instance of that spec will be returned.
+
+        If multiple root specs match the provided spec, or no root specs match
+        and multiple dependency specs match, then this raises an error
+        and reports all matching specs.
+        """
+        # Root specs will be keyed by concrete spec, value abstract
+        # Dependency-only specs will have value None
+        matches = {}
+
+        for user_spec, concretized_user_spec in self.concretized_specs():
+            if concretized_user_spec.satisfies(spec):
+                matches[concretized_user_spec] = user_spec
+            for dep_spec in concretized_user_spec.traverse(root=False):
+                if dep_spec.satisfies(spec):
+                    # Don't overwrite the abstract spec if present
+                    # If not present already, set to None
+                    matches[dep_spec] = matches.get(dep_spec)
+
+        if not matches:
+            return None
+        elif len(matches) == 1:
+            return list(matches.values())[0]
+
+        root_matches = dict((concrete, abstract)
+                            for concrete, abstract in matches.items()
+                            if abstract)
+
+        if len(root_matches) == 1:
+            return root_matches[0][1]
+
+        # More than one spec matched, and either multiple roots matched or
+        # none of the matches were roots
+        # If multiple root specs match, it is assumed that the abstract
+        # spec will most-succinctly summarize the difference between them
+        # (and the user can enter one of these to disambiguate)
+        root_fmt_str = 'Root spec %s\n  '
+        no_root_str = 'Dependency spec\n  '
+        match_strings = [
+            (root_fmt_str % abstract.format() if abstract else no_root_str) +
+            concrete.format('{hash:7}  ' + spack.spec.default_format)
+            for concrete, abstract in matches.items()
```

Comment:
I think this should be broken down some - I'm having difficulty reading this.

```
match_strings = []
for concrete, abstract in matches.items():
  if abstract:
    prefix = root_fmt_str % abstract.format()
  else:
    prefix = ''
  line = prefix + no_root_str + concrete.format()
  match_strings.append(line)
```

IMO it might make sense to omit the concrete spec entirely where an abstract spec is available (I assume the abstract specs must be distinct in the environment, so providing the abstract spec will be sufficient to distinguish it).

Reply:
I think we want the concrete specs in both cases, because we can only print hashes for concrete specs, and when abstract specs are very long, users will want to use the hashes to reference them. I'll break down the loop to make it more legible and then we can revisit this.

Reply:
That makes sense: showing the dependency spec in both cases sounds fine to me.